modules/dcache/src/main/java/org/dcache/services/ssh2/Ssh2Admin.java

```diff
@@ -318,6 +331,71 @@ public boolean authenticate(String userName, PublicKey key,
         }
     }
 
+    private static class RemoteHostValidator {
+
+        private enum Outcome {
+            ALLOW, DENY, DEFER
+        }
+
+        private boolean isValidHost(String line, ServerSession session) {
+            for (String linePart : Splitter.on(" ").trimResults().omitEmptyStrings().split(line)) {
+                if (linePart.startsWith("from=") && linePart.endsWith("\"")) {
+                    Set<Outcome> outcomes = EnumSet.noneOf(Outcome.class);
+                    for (String pattern : Splitter.on(",").trimResults().omitEmptyStrings().split(strip(linePart))) {
+                        outcomes.add(patternMatchesHost(pattern, session));
+                    }
+                    return outcomes.contains(Outcome.ALLOW) && !outcomes.contains(Outcome.DENY);
+                }
+            }
+            return true;
+        }
+
+        private Outcome patternMatchesHost(String pattern, ServerSession session) {
+            boolean patternMatches = false;
+            boolean patternIsNegated = pattern.startsWith("!");
+            if (patternIsNegated) {
+                pattern = pattern.substring(1);
+            }
+            SocketAddress remote = session.getClientAddress();
+            if (remote instanceof InetSocketAddress) {
+                String remoteAddress = ((InetSocketAddress) remote).getAddress().getHostAddress();
+                try {
+                    patternMatches = addressInCidrRange(pattern, remoteAddress);
+                } catch (IllegalArgumentException e) {
+                    String remoteName = ((InetSocketAddress) remote).getHostName();
+                    if (Glob.isGlob(pattern)) {
```

Comment:
I think the if-else clause is unnecessary: you can write this as just:

    Glob glob = new Glob(pattern);
    patternMatches = glob.matches(remoteName) || glob.matches(remoteAddress);

As is, there's a slight performance penalty if the pattern doesn't contain any wildcards, since the address would be tested twice, and there's (probably) some overhead in computing a regex state-machine instead of simply testing string equality.  I suspect that both cases don't add too much just now, and the latter should be fixed in Glob, anyway.